src/traces/box/select.js

```diff
@@ -0,0 +1,57 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var DESELECTDIM = require('../../constants/interactions').DESELECTDIM;
+
+module.exports = function selectPoints(searchInfo, polygon) {
+    var cd = searchInfo.cd;
+    var xa = searchInfo.xaxis;
+    var ya = searchInfo.yaxis;
+    var trace = cd[0].trace;
+    var node3 = cd[0].node3;
+    var selection = [];
+    var i, j;
+
+    if(trace.visible !== true) return [];
+
+    if(polygon === false) {
+        for(i = 0; i < cd.length; i++) {
+            for(j = 0; j < (cd[i].pts || []).length; j++) {
+                // clear selection
+                cd[i].pts[j].dim = 0;
+            }
+        }
+    } else {
+        for(i = 0; i < cd.length; i++) {
+            for(j = 0; j < (cd[i].pts || []).length; j++) {
+                var pt = cd[i].pts[j];
+                var x = xa.c2p(pt.x);
+                var y = ya.c2p(pt.y);
+
+                if(polygon.contains([x, y])) {
+                    selection.push({
+                        pointNumber: pt.i,
+                        x: pt.x,
+                        y: pt.y
+                    });
+                    pt.dim = 0;
+                } else {
+                    pt.dim = 1;
+                }
+            }
+        }
+    }
+
+    node3.selectAll('.point').style('opacity', function(d) {
+        return d.dim ? DESELECTDIM : 1;
```

Comment:
Note that this style update isn't compatible with the current trace attribute set.

To do so, we would need to add arrayOk support for `marker.color`, `marker.opacity` and `marker.symbol` would be pretty easy to add after https://github.com/plotly/plotly.js/pull/2094/commits/d352fa5d839fb428d57cab9f9886dc0370690c84

Reply:
... that is if no one finds too awkward to have both box-wide (e.g. `marker.outliercolor`) and per-sample-pt attributes in the same traces

Reply:
Seems reasonable. `outliercolor` is only for `boxpoints: 'suspectedoutliers'`, and in that mode selecting would be a bit weird. (side note: `outliercolor` should probably be described better in `attributes`... and perhaps the logic in `defaults` cleaned up or amended so we don't have the `outliercolor` attributes in `_fullData` if `boxpoints !== 'suspectedoutliers'`

Begs the question though of what selection (or crossfilter) should do in modes other than `boxpoints: 'all'` when at least the points within the box are hidden - what does selection do now? I could imagine it being cool to still be able to select the points, and have them pop into view when you do (likewise for crossfilter)... but that might be too complicated, dunno.

Reply:
> Begs the question though of what selection (or crossfilter) should do in modes other than boxpoints: 'all' when at least the points within the box are hidden - what does selection do now?

I'd vote for hiding the lasso and select mode bar buttons whenever `boxpoints !== 'all'` - similar to how we currently hide them from scatter-line-only graphs [here](https://github.com/plotly/plotly.js/blob/86e7135f54fedfa1e89b3e299833491369baa96e/src/components/modebar/manage.js#L178-L186). But maybe that's just because I'm lazy. cc @chriddyp @cpsievert any thoughts on this?

> and perhaps the logic in defaults cleaned up or amended so we don't have the outliercolor attributes in _fullData if boxpoints !== 'suspectedoutliers'

I believe that's [done](https://github.com/plotly/plotly.js/blob/86e7135f54fedfa1e89b3e299833491369baa96e/src/traces/box/defaults.js#L66-L69) already.

Reply:
> I'd vote for hiding the lasso and select mode bar buttons whenever `boxpoints !== 'all'`

We can certainly start that way anyhow, that's fine for this PR, we can revisit if there's interest later.

> I believe that's done already.

It's messy logic - but that's actually the [second coerce of one of them](https://github.com/plotly/plotly.js/blob/86e7135f54fedfa1e89b3e299833491369baa96e/src/traces/box/defaults.js#L49-L50) so the easy thing to do would be to delete them from the full trace if they're not needed. I guess the reason it's done this way is so you can set the default `boxpoints` to `'suspectedoutliers'` if either of the `outliercolor` values is a *real color* - stricter than just existing.

Reply:
Ha I see. I've been wanting to replace `Lib.coerce2` with `Lib.validate` (which doesn't fill in `fullData`). I'll do this in this PR :hammer:

Reply:
:hocho: `Lib.coerce2` turned out to be a little trickier than I thought - and want to do in this PR. In brief, we can _just_ replace `coerce2` by `validate`, `validate` return false whenever an value isn't set while `coerce2` returns the default value. 

So, here's a quick :hammer_and_wrench: and :lock: in https://github.com/plotly/plotly.js/pull/2094/commits/9ca410ac82a2c911c44d2fadd215103e157ed8aa

Reply:
> I'd vote for hiding the lasso and select mode bar buttons whenever boxpoints !== 'all'
We can certainly start that way anyhow, that's fine for this PR, we can revisit if there's interest later.

that's in https://github.com/plotly/plotly.js/pull/2094/commits/b4d80449eeb651eb69782540d51902cce18327c1